### ISSUE TITLE
phantom seems to hang after a while [ifort 2018 on ozstar]

### ISSUE THREAD

**Issue by cpinte** [danieljprice/phantom#3]
On ozstar, with module intel/2018.1.163-gcc-6.4.0 loaded, phantom appears to hang after a little bit. 

phantom was compiled with: 
"
export SYSTEM=ifort
make MCFOST=no MAXP=20000000 phantomsetup phantom
"
phantom hangs after 39 dumps (~ 1h) with the following setup file : 

"
# input file for disc setup routine

# resolution
                  np =     1000000    ! number of gas particles

# units
           dist_unit =          au    ! distance unit (e.g. au,pc,kpc,0.1pc)
           mass_unit =      solarm    ! mass unit (e.g. solarm,jupiterm,earthm)

# central object(s)/potential
            icentral =           1    ! use sink particles or external potential (0=potential,1=sinks)
              nsinks =           1    ! number of sinks

# options for central star
                  m1 =       2.000    ! star mass
               accr1 =       0.500    ! star accretion radius

# options for gas accretion disc
             isetgas =           0    ! how to set gas density profile (0=total disc mass,1=mass within annulus,2=surface density normalisation,3=surface density at reference radius,4=minimum Toomre Q)
           itapergas =           F    ! exponentially taper the outer disc profile
          ismoothgas =           F    ! smooth inner disc
               iwarp =           F    ! warp disc
                R_in =       1.000    ! inner radius
               R_ref =        100.    ! reference radius
               R_out =        100.    ! outer radius
              disc_m =       0.010    ! disc mass
              pindex =       0.000    ! p index
              qindex =       0.350    ! q index
             posangl =       0.000    ! position angle (deg)
                incl =       0.000    ! inclination (deg)
                 H_R =       0.100    ! H/R at R=R_ref
             alphaSS =       0.001    ! desired alphaSS

# set planets
          setplanets =           0    ! add planets? (0=no,1=yes)

# timestepping
             norbits =         100    ! maximum number of orbits at outer disc
              deltat =       0.001    ! output interval as fraction of orbital period
"

**Comment by danieljprice** [comment]
I can reproduce this, seems to hang 21 time steps after the last dump file

Running it with iverbose=5 gives the following:

```
> step 21 / 64 t = 174.7303 dt = 0.069 moved 78 in 4.9 cpu-s <
linear momentum error is  3.915E-10
angular momentum error is  1.845E-07
 
-------------------------------------------------
| bin |    dt    |   npart  |   frac  | cpufrac |
-------------------------------------------------
|  0    4.443E+00     931707   93.20%    0.00% |
|  1    2.221E+00      49590    4.96%    0.00% |
|  2    1.111E+00      13089    1.31%    7.67% |
|  3    5.554E-01       3704    0.37%    5.00% |
|  4    2.777E-01       1038    0.10%   13.06% |
|  5    1.388E-01        449    0.04%   25.14% |
|  6    6.942E-02         92    0.01%   49.14% |
-------------------------------------------------
> external/ptmass forces only : t=    174.738864
 dt(sink-sink) =   4.494232837155789E+307
           1 dt(ext/sink-sink) =   9.999999999999999E+028 , dt(sink-gas) =   8.567279732300316E-003
> external/ptmass forces only : t=    174.747431
 dt(sink-sink) =   4.494232837155789E+307
           2 dt(ext/sink-sink) =   9.999999999999999E+028 , dt(sink-gas) =   8.570752448573665E-003
> external/ptmass forces only : t=    174.756002
 dt(sink-sink) =   4.494232837155789E+307
           3 dt(ext/sink-sink) =   9.999999999999999E+028 , dt(sink-gas) =   8.574252806384351E-003
> external/ptmass forces only : t=    174.764576
 dt(sink-sink) =   4.494232837155789E+307
           4 dt(ext/sink-sink) =   9.999999999999999E+028 , dt(sink-gas) =   8.577777443585686E-003
> external/ptmass forces only : t=    174.773154
 dt(sink-sink) =   4.494232837155789E+307
           5 dt(ext/sink-sink) =   9.999999999999999E+028 , dt(sink-gas) =   8.581322856915074E-003
> external/ptmass forces only : t=    174.781735
 dt(sink-sink) =   4.494232837155789E+307
           6 dt(ext/sink-sink) =   9.999999999999999E+028 , dt(sink-gas) =   8.584885404014720E-003
> external/ptmass forces only : t=    174.790320
 dt(sink-sink) =   4.494232837155789E+307
Step: at time  1.748E+02,    2 particles were accreted amongst    1 sink(s).
           7 dt(ext/sink-sink) =   9.999999999999999E+028 , dt(sink-gas) =   8.588461306041385E-003
> external/ptmass forces only : t=    174.798908
 dt(sink-sink) =   4.494232837155789E+307
           8 dt(ext/sink-sink) =   9.999999999999999E+028 , dt(sink-gas) =   0.000000000000000E+000
```
which indicates the problem is that the sink-gas timestep goes to zero for some reason

**Comment by danieljprice** [comment]
Just a general comment, there should be no need to compile phantom with MAXP=, this is only needed for phantomsetup (and should not be necessary there either if you're using SETUP=disc)

**Comment by danieljprice** [comment]
update: I added an error check (376878ca), so now at least the hang is fixed (now gives a fatal error). So the symptoms are treated, but not yet the cause.

**Comment by danieljprice** [comment]
print*,'dtsinkgas =',dtsinkgas,C_force,1./sqrt(fonrmax),fonrmax,dtphi2

``` dtsinkgas =  9.999999999999999E+028  0.250000000000000       8.763509755138372E-147  1.302098537564868E+292  0.000000000000000E+000```

so the timestep is zero because dtphi2 = 0

**Comment by danieljprice** [comment]
apparently the mass of the sink particle (pmassj) is reported as near infinity, giving an almost infinite force and therefore a timestep that goes to zero:

 MASS=  8.404256815991542E+291  0.000000000000000E+000
 force= -7.334311771653479E+288  6.445157932094631E+288 -7.916096045600707E+287  3.344368081516690E+287

**Comment by danieljprice** [comment]
Happy to report this looks like an ifort bug, not a phantom bug. Issue was with openMP reduction operation on the "dptmass" array producing garbage. 

Workaround/fix implemented in a0b517c5, memory is now allocated explicitly instead of letting the compiler allocate memory automatically. This resolves the issue on ozstar.
